New Mobject: Added :class:`~.Tabular`

PhotonSpheres opened the pull request:
<!--
Thanks for your contribution to ManimCommunity!

Before filling in the details, ensure:
- Your local changes are up-to-date with ManimCommunity/manim
  
- The title of your PR gives a descriptive summary to end-users. Some examples:
  - Fixed last animations not running to completion
  - Added gradient support and documentation for SVG files
  Examples of what *NOT* to do:
  - "fixed that styling issue" - not descriptive enough
  - "fixed issue #XYZ" - end-user needs to do further research
-->
## Changelog / Overview
<!-- Optional (Recommended): a detailed overview of the PR for the upcoming
release's changelog entry. Useful for when the PR title isn't enough. 

DO NOT REMOVE THE FOLLOWING CHANGELOG LINES, EVEN IF YOU DON'T USE THEM.-->
<!--changelog-start-->
New Mobject that allows displaying multiple styles of tables.
<!--changelog-end-->

## Motivation
<!-- In what way do your changes improve the library? -->
Seems like a lot of people have their own version of a table. In my opinion, this could be used enough to implement it. I am open to hearing suggestions on whether this would be better as a plug-in. Also, feel free to comment on how a table should work. This is marked as a draft now since there is a lot still missing (especially docstrings).

## Explanation for Changes
<!-- How do your changes improve the library? -->
Allow a ready-to-go table logic for all sorts of numbers and mobjects. The implementation is very similar to `Matrix`, with the difference that the version here relies on `arrange_in_grid`.

<!-- For PRs introducing new features, please adjust
the link below to reference the docs of your feature. -->
[Tabular Landing page](https://manimce--1677.org.readthedocs.build/en/1677/reference/manim.mobject.tabular.html)
[Tabular](https://manimce--1677.org.readthedocs.build/en/1677/reference/manim.mobject.tabular.Tabular.html)
[DecimalTabular](https://manimce--1677.org.readthedocs.build/en/1677/reference/manim.mobject.tabular.DecimalTabular.html)
[IntegerTabular](https://manimce--1677.org.readthedocs.build/en/1677/reference/manim.mobject.tabular.IntegerTabular.html)
[MathTabular](https://manimce--1677.org.readthedocs.build/en/1677/reference/manim.mobject.tabular.MathTabular.html)
[MobjectTabular](https://manimce--1677.org.readthedocs.build/en/1677/reference/manim.mobject.tabular.MobjectTabular.html)
<!-- In the link above replace #### with you PR number.
You can also adjust the path to the module / class you worked on. This could look like adding
"/en/####/reference/manim.mobject.geometry.Arc.html" to the link for the Arc class.
Notice that the link will only work once the documentation is build which may take a few minutes.-->


## Testing Status
<!-- Optional (Recommended): your computer specs and what tests you ran with
their results, if any. This section is also intended for other
testing-related comments. -->
UPDATE: I wrote graphical unit tests for the module `tabular`.

If this goes forward I will write graphical tests for it.

## Further Comments
<!-- Optional: any further comments that might be useful for reviewers. -->
Check the documentation reference above for some examples.

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings
<!-- Once again, thanks for contributing to ManimCommunity! -->


<!-- Do not modify the lines below. These are for the reviewers of your PR -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

hydrobeam commented:
I just went through the docs and messed around with the tables and it looks pretty great! I think it does belong in the core of manim especially since the plugin environment isn't really well-developed. Just a couple comments so far:

- The typing should be implemented in the code itself, not the doc strings. As it was done in `Axes` for example:
https://github.com/ManimCommunity/manim/blob/main/manim/mobject/coordinate_systems.py#L962-L971 

- Maybe there should be an example which makes use of numpy's `reshape`? It can be pretty helpful for morphing data into the shape it needs to be in.

```py
class Tabletest(Scene):
    def construct(self):
        x_values = np.arange(1, 100).reshape(9,11)
        table = IntegerTabular(x_values, h_buff=0.5, v_buff=0.25)
        self.add(table)
```
- I'm not sure how feasible this is, but would it be possible to have the background rectangles span the entirety of the space taken up within the lines, as opposed to just the entry itself?
![image](https://user-images.githubusercontent.com/70682032/122657233-abbeac00-d12f-11eb-9456-f8fc78685ccb.png)

PhotonSpheres commented:
> I just went through the docs and messed around with the tables and it looks pretty great! I think it does belong in the core of manim especially since the plugin environment isn't really well-developed. Just a couple comments so far:

Cheers! And thank you for taking the time.

> * The typing should be implemented in the code itself, not the doc strings. As it was done in `Axes` for example:
>   https://github.com/ManimCommunity/manim/blob/main/manim/mobject/coordinate_systems.py#L962-L971

True, forgot about that. I tried to add typehints as well as I can. Someone with more knowledge can gladly correct me though.

> * Maybe there should be an example which makes use of numpy's `reshape`? It can be pretty helpful for morphing data into the shape it needs to be in.

Great idea! I added an example like this on the landing page.
 
>  * I'm not sure how feasible this is, but would it be possible to have the background rectangles span the entirety of the space taken up within the lines, as opposed to just the entry itself?
>   ![image](https://user-images.githubusercontent.com/70682032/122657233-abbeac00-d12f-11eb-9456-f8fc78685ccb.png)

I guess this would result in a complete background rectangle, so similar to set `include_background_rectangle=True`in the first place? Or am I misunderstanding this?

hydrobeam commented:
Yup that's the visual look that I had in mind, but the issue is that you can't select a specific square and adjust the colour (if you wanted to highlight something for example) with `include_background_rectangle = True`.  Ideally you could do something like
```py
table.get_entries()[5].background_rectangle.set_color(RED)
```
and be able to manually work with the rectangle for each entry

PhotonSpheres commented:
> ```python
> table.get_entries()[5].background_rectangle.set_color(RED)
> ```

This is a great idea, I implemented a `get_cell` and `get_highlighted_cell` method in `Tabular`, you can see examples for that in the docs now. Is this closer to what you had in mind?

hydrobeam commented:
I've adjusted the `arrange_in_grid` method in `OpenGLMobject` to make it compatible with this PR. I've tested it locally and it is able to be rendered now.

Darylgolden commented:
Is this PR ready for review? If so, can you mark it as such?

PhotonSpheres commented:
I think this is somewhat ready. For some reasons one of my graphic tests passes locally but fails for Ubuntu and Windows here (I have a Mac myself). I commented this test for now, maybe someone else can generate the test-data again from another machine to see if it works.

I am not comfortable with my typing/typehints. Someone with more knowledge should definitely check them. 

The whole `add_higlighted_cell` thing could maybe also be improved, not sure what would be best practice there.

Let me know what you think!

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+class IntegerTabular(Tabular):
+    """A mobject that displays a table with integer entries on the screen.
```
```suggestion
    """A specialized :class:`~.Tabular` mobject for use with with :class:`~.Integer`.
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+class DecimalTabular(Tabular):
+    """A mobject that displays a table with decimal entries on the screen.
```
```suggestion
    """A specialized :class:`~.Tabular` mobject for use with with :class:`~.DecimalNumber` to display decimal entries.
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+        **kwargs,
+    ):
+        """
+        Special case of :class:`~.Tabular` with `element_to_mobject` set to :class:`~.DecimalNumber`.
```
```suggestion
        Special case of :class:`~.Tabular` with ``element_to_mobject`` set to :class:`~.DecimalNumber`.
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+    ):
+        """
+        Special case of :class:`~.Tabular` with `element_to_mobject` set to :class:`~.DecimalNumber`.
+        By default, `num_decimal_places` is set to 1.
+        Will round/truncate the decimal places as per the provided config.
```
```suggestion
        By default, ``num_decimal_places`` is set to 1.
        Will round/truncate the decimal places based on the provided ``element_to_mobject_config``.
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+            else:
+                raise ValueError("Not all rows in table have the same length.")
+        VGroup.__init__(self, **kwargs)
+        mob_table = self.table_to_mob_table(table)
```
It'd be helpful if there were some comments/organization to make it easier to parse what's happening here.

PhotonSpheres replied:
You mean in total or a specific line you have in mind?

hydrobeam replied:
I think now that the private methods are documented this isn't necessary 👍

hydrobeam reviewed (changes requested):
Docs comments + semi-minor code adjustments.

PhotonSpheres commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+    def __init__(
+        self,
+        table: Iterable[Iterable[Union[float, str, "VMobject"]]],
+        row_labels: Optional[Iterable["VMobject"]] = None,
+        col_labels: Optional[Iterable["VMobject"]] = None,
```
My thought process is that you can do that by just don't use any labels at all, at least if your entries and and your are both of same type. But I understand where you are coming from.

PhotonSpheres commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+
+        class MobjectTabularExample(Scene):
+            def construct(self):
+                cross = VGroup(
+                    Line(UP + LEFT, DOWN + RIGHT),
+                    Line(UP + RIGHT, DOWN + LEFT),
+                )
```
I was looking at that, but that Mobject needs another Mobject as input which I found weird for this use case here, I think the predefined cross is more to "cross" something out?

hydrobeam replied:
Ah that's weird, I didn't notice that. Probably needs a rework 😅

PhotonSpheres commented on tests/test_graphical_units/test_tabular.py:
```diff
@@ -0,0 +1,54 @@
+from ..utils.GraphicalUnitTester import GraphicalUnitTester
+from ..utils.testing_utils import get_scenes_to_test
+
+# class TabularTest(Scene):
+#     def construct(self):
+#         t = Tabular(
+#             [["1", "2"], ["3", "4"]],
+#             row_labels=[Text("R1"), Text("R2")],
+#             col_labels=[Text("C1"), Text("C2")],
+#             top_left_entry=Text("TOP"),
+#             include_outer_lines=True,
+#         )
+#         self.add(t)
```
Good catch, cheers!

PhotonSpheres commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+                pass
+            else:
+                raise ValueError("Not all rows in table have the same length.")
+        VGroup.__init__(self, **kwargs)
```
If I do that, all checks fail. No idea why

hydrobeam commented:
`super()` wasn't working because you forgot to remove the `self` from the `__init__` which isn't necessary with `super`

PhotonSpheres commented:
During my changes I got this conflict in the `docs/source/reference.rst` file but I don't know why. I just added the `tabular` file from the beginning and it always worked but now it doesn't? Anyone any ideas?

PhotonSpheres commented:
> EDIT: Also, I noticed that some of the suggestions were resolved without being implemented. Was this an error or did you disagree with them?

I hope I didn't do that? I just implemented some of those directly on my local machine and pushed one last commit because the commits via Github were too much. I think I implemented everything I resolved. Did I miss something?

PhotonSpheres commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+                        ["Third","Fourth"]],
+                        row_labels=[Text("R1"), Text("R2")],
+                        col_labels=[Text("C1"), Text("C2")])
+                    table.add_highlighted_cell((2,2), color=GREEN)
```
Since you add the highlighted cell to the tabular mobject I guess the `add` makes sense now, no?

PhotonSpheres commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1089 @@
+                label_group.add(*label)
+        return label_group
+
+    def add_background_to_entries(self) -> "Tabular":
```
Implemented!

hydrobeam replied:
Was it 😅? I noticed a colour commit for the class (which is useful!) but I meant for this method specifically too. like being able to pass in `color = GREEN`, with a default of BLACK

hydrobeam replied:
I think you may have accidentally resolved this 🤔

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+                    top_left_entry=Star().scale(0.3),
+                    include_outer_lines=True,
+                    line_config={"stroke_width": 1, "color": YELLOW})
+                t3.remove(*[line for line in t3.get_vertical_lines()])
```
**(suggestion)**: a vgroup can be unpacked like a list.
```suggestion
                t3.remove(*t3.get_vertical_lines())
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+        background_rectangle_color
+            Background color of table if ``include_background_rectangle`` is ``True``.
+        element_to_mobject
+            The :class:`~.Mobject` class applied to the table entries. by default :class:`~.Paragraph`. For common choices, see :mod:`~.text_mobject`/:mod`~.tex_mobject`.
```
**(suggestion)**: typo
```suggestion
            The :class:`~.Mobject` class applied to the table entries. by default :class:`~.Paragraph`. For common choices, see :mod:`~.text_mobject`/:mod:`~.tex_mobject`.
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+
+        Returns
+        --------
+        List[:class:`~.VGroup`]
```
**(comment/suggestion)**: it actually returns a `VGroup` of vgroups. not a list of vgroups,  VGroup[VGroup] might be valid here.

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+        )
+
+    def get_rows(self) -> List[VGroup]:
+        """Return the rows of the table as a list of :class:`~.VGroup`.
```
Same as above.

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+        label_group = VGroup()
+        if self.top_left_entry is not None:
+            label_group.add(self.top_left_entry)
+        for label in [self.col_labels, self.row_labels]:
```
**(suggestion, nitpick)**: a tuple would be more appropriate here since you're not changing the values
```suggestion
        for label in (self.col_labels, self.row_labels):
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+            A 2D array, or a list of lists. Content of the table must be valid input
+            for :class:`~.DecimalNumber`.
+        element_to_mobject
+            Element to mobject, here set as :class:`~.DecimalNumber`.
```
**(suggestion)**: Missed these last two the last time around.
```suggestion
            The :class:`~.Mobject` class applied to the table entries. Set as :class:`~.DecimalNumber`.
```

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+            A 2d array or list of lists. Content of the table has to be valid input
+            for :class:`~.Integer`.
+        element_to_mobject
+            Element to mobject, here set as :class:`~.Integer`.
```
```suggestion
            The :class:`~.Mobject` class applied to the table entries. Set as :class:`~.Integer`.
```

hydrobeam reviewed (changes requested):
This PR mostly LGTM, but I've left a few comments/suggestions throughout. 

Also, I tried adopting some standards from https://conventionalcomments.org/ for this review, hopefully it was a little more helpful 😅.

PhotonSpheres commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1114 @@
+.. manim:: TabularExamples
+    :save_last_frame:
+
+    class TabularExamples(Scene):
```
Wow, this does look wonky...

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1122 @@
+                    col_labels = [self.top_left_entry] + self.col_labels
+                    mob_table.insert(0, col_labels)
+                else:
+                    # Placeholder to use arrange_in_grid id top_left_entry is not set.
```
```suggestion
                    # Placeholder to use arrange_in_grid if top_left_entry is not set.
```

hydrobeam reviewed (approved):
I think this PR is good to go! Thank you for this great addition to the library (and the amazing docs to go along with it!) 

Just add the `color` parameter to `add_background_to_entries` +  I've added a quick suggestion for a typo 👍

hydrobeam commented on manim/mobject/tabular.py:
```diff
@@ -0,0 +1,1121 @@
+                label_group.add(*label)
+        return label_group
+
+    def add_background_to_entries(self) -> "Tabular":
+        """Adds a black :class:`~.BackgroundRectangle` to each entry of the table."""
+        for mob in self.get_entries():
+            mob.add_background_rectangle(color=self.entries_background_color)
+        return self
```
I was thinking more along the lines of passing in a paramter, sorry for not being more explicit 🙇

The code in the `__init__` would need to be changed too 😅  but github doesn't make it easy to do suggestions like that.

```suggestion
    def add_background_to_entries(self, color: Color = BLACK) -> "Tabular":
        """Adds a :class:`~.BackgroundRectangle` to each entry of the table."""
        for mob in self.get_entries():
            mob.add_background_rectangle(color=color)
        return self
```

PhotonSpheres replied:
Yeah I did that in a stupid way, should be fixed now!
